R/stan_betareg.R

```diff
@@ -0,0 +1,198 @@
+#' z <- rnorm(N, 2, 1)
+#' mu <- binomial(link = "logit")$linkinv(1 + 0.2*x)
+#' phi <- exp(1.5 + 0.4*z)
+#' y <- rbeta(N, mu * phi, (1 - mu) * phi)
+#' hist(y, col = "dark grey", border = FALSE, xlim = c(0,1))
+#' fake_dat <- data.frame(y, x, z)
+#' 
+#' fit <- stan_betareg(y ~ x | z + x, data = fake_dat, 
+#'                     link = "logit", link.phi = "log")
+#' print(fit, digits = 2)
+#' plot(fit)
+#' pp_check(fit)
+#' }
+#' 
+stan_betareg <- function(formula, data, subset, na.action, weights, offset,
+                         link = c("logit", "probit", "cloglog", "cauchit", "log", "loglog"),
+                         link.phi = c("log", "identity", "sqrt"), 
+                         model = TRUE, y = TRUE, x = FALSE, ...,
+                         prior = normal(), prior_intercept = normal(),
+                         prior_z = normal(), prior_intercept_z = normal(),
+                         prior_ops = prior_options(), prior_PD = FALSE, 
+                         algorithm = c("sampling", "optimizing", "meanfield", "fullrank"),
+                         adapt_delta = NULL, QR = FALSE, sparse = FALSE) {
+  
+  if (!requireNamespace("betareg", quietly = TRUE)) 
+    stop("Please install the betareg package before using 'stan_betareg'.")
+  
+  mc <- match.call(expand.dots = FALSE)
+  mc$model <- mc$y <- mc$x <- TRUE
+  
+  # NULLify any Stan specific arguments in mc now
+  mc$prior <- mc$prior_intercept <- mc$prior_ops <- mc$prior_PD <- mc$algorithm <-
+    mc$adapt_delta <- mc$QR <- mc$sparse <- NULL
+  
+  mc$drop.unused.levels <- TRUE
+  mc[[1L]] <- quote(betareg::betareg)
+  mc$control <- betareg::betareg.control(maxit = 0, fsmaxit = 0)
+  br <- suppressWarnings(eval(mc, parent.frame()))
+  mf <- check_constant_vars(br$model)
+  mt <- br$terms
+  Y <- array1D_check(model.response(mf, type = "any"))
+  X <- model.matrix(br)
+  Z <- model.matrix(br, model = "precision")
+  weights <- validate_weights(as.vector(model.weights(mf)))
+  offset <- validate_offset(as.vector(model.offset(mf)), y = Y)
+  if (!length(prior_ops)) 
+    prior_ops <- list(scaled = FALSE, prior_scale_for_dispersion = Inf)
+  
+  # determine whether user specified regression matrix for precision model
+  if (length(grep("\\|", all.names(formula))) == 1 && length(link.phi) > 1) {
+    link.phi <- "log"
+  }
+  
+  # pass the prior information to stan_betareg.fit()
+  stanfit <- stan_betareg.fit(x = X, y = Y, z = Z, 
+                              weights = weights, offset = offset, 
+                              link = link, link.phi = link.phi, ..., 
+                              prior = prior, prior_z = prior_z, 
+                              prior_intercept = prior_intercept, 
+                              prior_intercept_z = prior_intercept_z,
+                              prior_ops = prior_ops, prior_PD = prior_PD, 
+                              algorithm = algorithm, adapt_delta = adapt_delta,
+                              QR = QR, sparse = FALSE)
```

Comment:
@imadmali Also just wanted to check again about why `sparse` is fixed to `FALSE` here